### PR TITLE
Add `{Dispatcher, RunNow, System}::dispose`

### LINKED ISSUE #119: Add `System::dispose`
Requested by @Frizi.

`System::dispose` would be called by `Dispatcher::dispose` and can be used to clean up externally allocated resources.

## Unresolved

* Signature: `fn dispose(&self, world: &World)` / `fn dispose(&self, world: &mut World)` / `fn dispose(&mut self, world: &World)` / `fn dispose(&mut self, world: &mut World)`
* Is it mandatory to call it? How to make sure it gets called?

### PR DESCRIPTION
**Pull request by torkleyy**
Fixes #119

### CODE REVIEW

File: src/system.rs [hunk 1]
```diff
@@ -121,6 +121,11 @@ pub trait RunNow<'a> {
 
     /// Sets up `World` for a later call to `run_now`.
     fn setup(&mut self, world: &mut World);
+
+    /// Performs clean up that requires resources from the `World`.
+    /// This commonly removes components from `world` which depend on external
+    /// resources.
+    fn dispose(&mut self, world: &mut World);
```
Comment (LucioFranco) [inline src/system.rs]: prob want a default impl here and an impl in the blanket? This should prob also reduce the amount of breaking thatll happen. I assume its safe to just do nothing here as a default?

File: src/system.rs [hunk 2]
```diff
@@ -185,6 +194,13 @@ pub trait System<'a> {
     fn setup(&mut self, world: &mut World) {
         <Self::SystemData as DynamicSystemData>::setup(&self.accessor(), world)
     }
+
+    /// Performs clean up that requires resources from the `World`.
+    /// This commonly removes components from `world` which depend on external
+    /// resources.
+    fn dispose(&mut self, world: &mut World) {
+        let _ = world;
```
Comment (LucioFranco) [inline src/system.rs]: do you actually need to do this?

Comment (torkleyy) [reply src/system.rs]: If I want no warning and keep `world` (instead of `_world`) as a parameter, yes :)

Comment (LucioFranco) [reply src/system.rs]: why not just `_world`?

File: src/system.rs (hunk 1 above)
Comment (torkleyy) [reply src/system.rs]: I'm not sure. The `RunNow` API is pretty much for consuming only, not for implementing.

File: src/system.rs (hunk 2 above)
Comment (torkleyy) [reply src/system.rs]: It shows up in the docs ^^

Comment (LucioFranco) [reply src/system.rs]: but blanket impls you dont care about the fn's generally you just care about that the bounds for T work for you no? I mean this is minor i dont mind either way 😅

Comment (torkleyy) [reply src/system.rs]: This is not a blanket impl, it's the `System` trait. I just wanted to have a nice parameter show up in the docs :sweat_smile:

Comment (LucioFranco) [reply src/system.rs]: Oh my god! You're right I missed that haha, I guess I didnt expand the diff hahaha. you're right this works! Sorry!

**Comment by torkleyy** [comment]
bors delegate=Frizi

**Comment by Frizi** [comment]
Haven't had a chance to review it before, but it does look very good :) Just a question. Why is `dispose` setting a flag instead of consuming the dispatcher? I think it's easier to work with that way, but it's also not entirely safe.

**Comment by torkleyy** [comment]
Thanks for taking a look! That's a good idea, yeah. Totally missed that.

**Pull request merged**